gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/MysqlDagStateStoreV2.java

```diff
@@ -0,0 +1,252 @@
+      + "PRIMARY KEY (dag_node_id))";
+
+  protected static final String INSERT_STATEMENT = "INSERT INTO %s (dag_node_id, parent_dag_id, dag_node) "
+      + "VALUES (?, ?, ?) AS new ON DUPLICATE KEY UPDATE dag_node = new.dag_node";
+  protected static final String GET_DAG_STATEMENT = "SELECT dag_node FROM %s WHERE parent_dag_id = ?";
+  protected static final String GET_DAG_NODE_STATEMENT = "SELECT dag_node FROM %s WHERE dag_node_id = ?";
+  protected static final String DELETE_DAG_STATEMENT = "DELETE FROM %s WHERE parent_dag_id = ?";
+  protected static final String DELETE_DAG_NODE_STATEMENT = "DELETE FROM %s WHERE dag_node_id = ?";
+  private final ContextAwareCounter totalDagCount;
+
+  public MysqlDagStateStoreV2(Config config, Map<URI, TopologySpec> topologySpecMap) throws IOException {
+    if (config.hasPath(CONFIG_PREFIX)) {
+      config = config.getConfig(CONFIG_PREFIX).withFallback(config);
+    }
+
+    String DEFAULT_TABLE_NAME = "dag_node_state_store";
+    this.tableName = ConfigUtils.getString(config, ConfigurationKeys.STATE_STORE_DB_TABLE_KEY, DEFAULT_TABLE_NAME);
+    // create table if it does not exist
+    DataSource dataSource = MysqlDataSourceFactory.get(config, SharedResourcesBrokerFactory.getImplicitBroker());
+
+    try (Connection connection = dataSource.getConnection();
+        PreparedStatement createStatement = connection.prepareStatement(String.format(CREATE_TABLE_STATEMENT, tableName))) {
+      createStatement.executeUpdate();
+      connection.commit();
+    } catch (SQLException e) {
+      throw new IOException("Failure creation table " + tableName, e);
+    }
+    this.dbStatementExecutor = new DBStatementExecutor(dataSource, log);
+
+    JsonSerializer<List<JobExecutionPlan>> serializer = new JobExecutionPlanListSerializer();
+    JsonDeserializer<List<JobExecutionPlan>> deserializer = new JobExecutionPlanListDeserializer(topologySpecMap);
+    Type typeToken = new TypeToken<List<JobExecutionPlan>>() {
+    }.getType();
+    this.serDe = new GsonSerDe<>(typeToken, serializer, deserializer);
+    this.jobExecPlanDagFactory = new JobExecutionPlanDagFactory();
+    MetricContext metricContext =
+        Instrumented.getMetricContext(new State(ConfigUtils.configToProperties(config)), this.getClass());
+    this.totalDagCount = metricContext.contextAwareCounter(ServiceMetricNames.DAG_COUNT_MYSQL_DAG_STATE_COUNT);
+
+  }
+
+  @Override
+  public void writeCheckpoint(Dag<JobExecutionPlan> dag)
+      throws IOException {
+    DagManager.DagId dagId = DagManagerUtils.generateDagId(dag);
+    boolean newDag = false;
+    for (Dag.DagNode<JobExecutionPlan> dagNode : dag.getNodes()) {
+      if (addDagNodeState(dagNode, dagId) == 1) {
+        newDag = true;
+      }
+    }
+    if (newDag) {
+      this.totalDagCount.inc();
+    }
+  }
+
+  @Override
+  public void cleanUp(Dag<JobExecutionPlan> dag) throws IOException {
+    cleanUp(generateDagId(dag).toString());
```

Comment:
won't this call the version throwing`NotSupportedException`?

Reply:
corrected